### PR TITLE
Feature: git stash in command panel

### LINKED ISSUE #628: Commands to manipulate the stash
There are no commands in the palette to pop and add to the stash.
Any objection to adding those?

### PR DESCRIPTION
**Pull request by stoivo**
fix #628

-----

The maintainers of this repo require that all pull request submitters agree and adhere to the following:

- [x] I have read and agree to the [contribution guidelines](https://github.com/divmain/GitSavvy/blob/master/CONTRIBUTING.md).
 _(required)_
- [x] All related documentation has been updated to reflect the changes made. _(required)_
- [x] My commit messages are cleaned up and ready to merge. _(required)_

The maintainers of this repository require you to select the semantic version type that
the changes in this pull request represent.  Please select one of the following:
- [ ]  major
- [ ]  minor
- [x]  patch
- [ ]  documentation only

### CODE REVIEW

**Comment by randy3k** [comment]
should we call the commands `git: stash xxxx` instead of `git: xxx stash`? Just to match the regular command line input order.

**Comment by randy3k** [comment]
We should also add some docs for `git: stash`. And we forgot the docs for `git: revert` too.

File: core/ui_mixins/quick_panel.py [hunk 1]
```diff
@@ -500,3 +500,30 @@ def on_selection(self, commit):
     def on_selection_async(self, commit):
         sublime.active_window().run_command("hide_panel", {"panel": "output.show_commit_info"})
         self.on_done(commit)
+
+
+def show_stash_panel(entries, on_done):
```
Comment (randy3k) [inline core/ui_mixins/quick_panel.py]: Do we really need a new function to show stash if `entries` are to be provided?  `show_paginated_panel` should just work. If you really want to implement a panel for stashes. Then it should be something similar to `show_branch_panel` which only takes the `on_done` argument.

**Comment by stoivo** [comment]
> should we call the commands git: stash xxxx instead of git: xxx stash? Just to match the regular command line input order.

> Edit: but it is a bit strange to have git: stash show and git: stash create...

git: xxx stash reads more natural but it is the wrong order compared to the command line. 

> We should also add some docs for git: stash. And we forgot the docs for git: revert too.

I can add docs for both tomorrow.

File: core/ui_mixins/quick_panel.py (hunk 1 above)
Comment (stoivo) [inline core/ui_mixins/quick_panel.py]: I think is it better to have this function than specifying everything each time.  Maybe StashPanel 
 https://github.com/divmain/GitSavvy/pull/744/files#diff-6809e629d1384bcdee2aa618a5425cbfR517  is an overkill. I just copied the above one and removed all we don't need.

Comment (randy3k) [inline core/ui_mixins/quick_panel.py]: Then the entries should be generated in the `show` method of `StashPanel` as in the [`show_remote_panel`](https://github.com/divmain/GitSavvy/blob/cf2b6e3992a93b5bbae1d3d8b963d327949a1ea6/core/ui_mixins/quick_panel.py#L140)

**Comment by stoivo** [comment]
Please review again. @randy3k

**Comment by randy3k** [comment]
added a few more commits, mostly for the StashPanel.

**Comment by randy3k** [comment]
I am still struggle between `git: create stash` and `git: stash create`. 

@aristidesfl since you requested it, do you have any thoughts?

**Comment by randy3k** [comment]
@stoivo please free to merge it if 👌

File: core/ui_mixins/quick_panel.py [hunk 2]
```diff
@@ -502,15 +502,22 @@ def on_selection_async(self, commit):
...
-        on_done,
-        format_item=(lambda entry: entry.id + " " + entry.description))
-    lp.show()
-    return lp
+    sp = StashPanel(on_done, **kwargs)
+    sp.show()
+    return sp
+
+
+class StashPanel(PaginatedPanel, GitCommand):
```
Comment (stoivo) [inline core/ui_mixins/quick_panel.py]: I thought you wanted to avoid creating this class

Comment (randy3k) [reply core/ui_mixins/quick_panel.py]: Oh, I meant either creating  `StashPanel` which computes stashes directly, or just use `show_paginated_panel`. But you were in between, your `StashPanel` still needs the list of stashes as an argument.

**Comment by aristidesfl** [comment]
@randy3k I think it is simpler to follow the command line order. There are more benefits to it, than otherwise. Also the command line uses `git stash save`, not `git stash save`.

Have you thought about also providing `git: stash`  as an alias to `git stash save`?

**Comment by randy3k** [comment]
Ya..you are right about the save option. I also agree it should be consistent with the command line. 

I don't think we should alias `git: stash`, but rather it will pop up a quickpanel of all the available suboptions.

**Comment by randy3k** [comment]
Ya..you are right about the save option. I also agree they should be consistent with the command line. 

But I don't think we should alias `git: stash`, but rather it will pop up a quickpanel of all the available suboptions.

**Comment by stoivo** [comment]
Also the behavur could be better when you have no stashed.

**Comment by aristidesfl** [comment]
So all stash options will be available via a sub-menu? 
Stash save also supports an optional description parameter: `git stash save 'description'`.
Does the PR support that?

**Comment by randy3k** [comment]
They are now available as `git: stash save`, `git stash pop` etc. You will be asked a description when you create a stash.

**Comment by randy3k** [comment]
👍 ? @stoivo

**Comment by stoivo** [comment]
I would like to improve the behaviour if there no stashes.

**Comment by randy3k** [comment]
We could extend PaginatedPanel by adding an argument like `empty_message` along with `next_message`

**Comment by stoivo** [comment]
I did not want to change the defaut bevavour. But i think that would make sense. Can you commit the changes you thought of?

File: core/ui_mixins/quick_panel.py [hunk 3]
```diff
@@ -460,33 +468,31 @@ def is_done(self):
         return self._is_done
 
 
-def show_log_panel(entries, on_done, limit=6000, selected_index=None, on_highlight=None):
+def show_log_panel(entries, on_done, **kwargs):
```
Comment (stoivo) [inline core/ui_mixins/quick_panel.py]: You really like `**kwargs` :D
This looks good.

**Comment by stoivo** [comment]
Merge?

**Comment by randy3k** [comment]
Let 's merge and test it in the dev branch.

**Pull request merged**